python/cugraph/cugraph/dask/structure/mg_property_graph.py

```diff
@@ -195,6 +166,33 @@ def edge_property_names(self):
             return props
         return []
 
+    @property
```

Comment:
Apologies for not bringing this up sooner, but I'm now wondering if we can take advantage of [@cached_property](https://docs.python.org/3/library/functools.html?highlight=cached_property#functools.cached_property) here and elsewhere.  In this case, whenever `add_vertex_data` is called, the `vertex_types` attr would be deleted to cause this code to run and cache on the new data.

The downside is that @cached_property allows writes, but there's an easy and fairly clean workaround for that.
Thoughts?
```python
>>> from functools import cached_property
>>> class Foo:
...    some_expensive_computation_result = 33
...    @cached_property
...    def __foo(self):
...       print("running")
...       return self.some_expensive_computation_result
...    @property
...    def foo(self):
...       return self.__foo
...    def update_data(self):
...       del self.__foo
...
>>> f=Foo()
>>> f.foo
running
33
>>> f.foo
33
>>> f.foo=22
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: can't set attribute
>>> f.update_data()
>>> f.foo
running
33
>>> f.foo
33
```

Reply:
Good question.  I would only use `@cached_property` for truly immutable results.  Since we need to clear the property, I don't think using `@cached_property` buys us anything, since it's still complex.  I also prefer to avoid adding or removing keys to an objects `__dict__` so that we get the advantages of [PEP 412](https://peps.python.org/pep-0412/).

Perhaps we should try to be more consistent with how we cache properties.

Reply:
You convinced me with PEP 412 and the fact that we have to `del self.__prop` to clear it which would constantly change the instance keys.

I do think that we should eventually (soon?) look into a general caching pattern of some kind for PG (and possibly others), since the usage pattern seems to be 1. add data in one or several add steps, 2. access data often. We do support adding data after accessing it of course, but I wouldn't want to penalize the most common case (at least as we've seen so far for GNNs) by not having any caching at all.